java-engine/src/main/java/io/getunleash/engine/UnleashEngine.java

```diff
@@ -251,50 +227,83 @@ public VariantDef getVariant(String toggleName, Context context)
 
-    Variant variant = Variant.getRootAsVariant(buf);
-    if (variant.name() == null) {
-      dealloc.apply(contextPtr, contextBytes.length);
-      deallocResponseBuffer.apply(ptr, len);
-      return null;
-    }
-    VariantPayload messagePayload = variant.payload();
-    Payload payload = null;
+      if (variantPayload != null) {
+        payload = new Payload();
+        payload.setType(variant.payload().payloadType());
+        payload.setValue(variant.payload().value());
+      }
 
-    if (messagePayload != null) {
-      payload = new Payload();
-      payload.setType(messagePayload.payloadType());
-      payload.setValue(messagePayload.value());
+      return new VariantDef(variant.name(), payload, variant.enabled(), variant.featureEnabled());
+    } else {
+      return null;
     }
-
-    dealloc.apply(contextPtr, contextBytes.length);
-    deallocResponseBuffer.apply(ptr, len);
-
-    return new VariantDef(variant.name(), payload, variant.enabled(), variant.featureEnabled());
   }
 
   public MetricsBucket getMetrics() throws JsonMappingException, JsonProcessingException {
     ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
+
     long packed = (long) getMetrics.apply(this.enginePointer, now.toInstant().toEpochMilli())[0];
+    MetricsBucket bucket = derefWasmPointer(packed, MetricsBucket::getRootAsMetricsBucket);
+
+    if (bucket.togglesVector() == null) {
+      return null;
+    }
+    return bucket;
+  }
+
+  private <T> T derefWasmPointer(long packed, Function<ByteBuffer, T> decoder) {
+    // This is not sane. To receive the response we need to two things:
+    // 1) a pointer
+    // 2) a length so we can read the pointer value to the end but not beyond
+    //
+    // However, we don't have a way to pass complex objects back to the host
+    // function. We can use a pre-allocated shared buffer but we would need to have
+    // that buffer size appropriately tuned for real workloads.
+    // Which requires a bunch of experimentation sooooo...
+    // instead we hack this. We're using 32 bit WASM here, which means
+    // pointers are 32 bits and we need a second 32 bit number to represent the
+    // length of the buffer. We can pass a 64 bit number across the WASM boundary,
+    // which is really two 32 bit numbers wearing a silly hat
+
     int ptr = (int) (packed & 0xFFFFFFFFL);
     int len = (int) (packed >>> 32);
+
     byte[] bytes = instance.memory().readBytes(ptr, len);
 
     ByteBuffer buf = ByteBuffer.wrap(bytes);
-    buf.order(ByteOrder.LITTLE_ENDIAN); // Apparently flatBuffers are little-endian
-    MetricsBucket bucket = MetricsBucket.getRootAsMetricsBucket(buf);
+    buf.order(ByteOrder.LITTLE_ENDIAN);
+
+    T response = decoder.apply(buf);
     deallocResponseBuffer.apply(ptr, len);
-    if (bucket.togglesVector() == null) {
-      return null;
-    }
-    return bucket;
+    return response;
+  }
+
+  private <T> T callWasmFunctionWithResponse(
```

Comment:
The important part - now this is the primary way to call into the WASM Ygg layer rather then rewriting it each time